### ISSUE TITLE
plasma store full (linux)

### ISSUE THREAD

**Issue by pchalasani** [ray-project/ray#1175]
Context -- I'm launching remote jobs to read files from s3 into pandas dataframes. 

before I delve into details, I first wanted to check if there's a simple idea/suggestion that will fix the issue:

I'm getting a "plasma store full" error
```
24 | file_num, row_pkl = ray.get( ready_ids[0] )
25 | File "/usr/local/lib/python3.5/dist-packages/ray/worker.py", line 2058, in get
26 | raise RayGetError(object_ids, value)
27 | ray.worker.RayGetError: Could not get objectid ObjectID(2d19bc6527ce7f4aff1e8c497eef5ba4991c83bb). It was created by remote function [31mseqdata.eventdata.read_file[39m which failed with:
28 |  
29 | Remote function [31mseqdata.eventdata.read_file[39m failed with:
30 |  
31 | Traceback (most recent call last):
32 | File "/usr/local/lib/python3.5/dist-packages/ray/worker.py", line 747, in _process_task
33 | self._store_outputs_in_objstore(return_object_ids, outputs)
34 | File "/usr/local/lib/python3.5/dist-packages/ray/worker.py", line 670, in _store_outputs_in_objstore
35 | self.put_object(objectids[i], outputs[i])
36 | File "/usr/local/lib/python3.5/dist-packages/ray/worker.py", line 330, in put_object
37 | self.store_and_register(object_id, value)
38 | File "/usr/local/lib/python3.5/dist-packages/ray/worker.py", line 283, in store_and_register
39 | object_id.id()), self.serialization_context)
40 | File "pyarrow/plasma.pyx", line 395, in pyarrow.plasma.PlasmaClient.put (/ray/src/thirdparty/arrow/python/build/temp.linux-x86_64-3.5/plasma.cxx:5130)
41 | File "pyarrow/plasma.pyx", line 301, in pyarrow.plasma.PlasmaClient.create (/ray/src/thirdparty/arrow/python/build/temp.linux-x86_64-3.5/plasma.cxx:4382)
42 | File "pyarrow/error.pxi", line 93, in pyarrow.lib.check_status (/ray/src/thirdparty/arrow/python/build/temp.linux-x86_64-3.5/lib.cxx:8476)
43 | pyarrow.lib.PlasmaStoreFull: object does not fit in the plasma store
```

**Comment by robertnishihara** [comment]
When Ray starts up, it prints out how much memory it reserves for the object store, e.g.,

```
Allowing the Plasma store to use up to 13.7439GB of memory.
```

On Linux this should be about 40% of the total memory. What is this number in your case? And how much memory do you expect your application to need?

Normally when the object store fills up, it tries to evict some objects. You can see this behavior by running a loop like

```python
import numpy as np
import ray

ray.init()

x = np.random.normal(size=1000000)

while True:
    ray.put(x)
```

Eventually it will print something like

```
There is not enough space to create this object, so evicting 344 objects to free up 2752305472 bytes.
```

If it isn't evicting stuff in your case, that could mean one of the following:

1) You're trying to create a single object that is so large it can't fit in the object store.
2) Many of the objects in the object store are still in use by some worker so they can't be evicted. This should only happen if numpy arrays are involved.

**Comment by pchalasani** [comment]
At the start, it doesn't print the line you said, but it does complain about having only 67MB. I'm running this on a remote GCP GPU via neptune.ml (not in a docker container). 
```
Waiting for redis server at 127.0.0.1:12260 to respond...
--
176 | Waiting for redis server at 127.0.0.1:13186 to respond...
177 | Warning: Reducing object store memory because /dev/shm has only 67108864 bytes available. You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you may need to pass an argument with the flag '--shm-size' to 'docker run'.
178 | Starting local scheduler with 16 CPUs, 1 GPUs
```

And I also see in stdout some messages about eviction:

```

ray.init(redis_address="127.0.0.1:12260")
--
341 |  
342 | An object created by ray.put was evicted and could not be reconstructed. The driver may need to be restarted.
```

**Comment by robertnishihara** [comment]
Does 67MB sound right? Or do you expect there to be more memory. Note that sometimes there can be plenty of memory but the size of the *shared memory file system* is too small.

Also, can you share a simple script that reproduces the problem? It looks like the problem is just that there is not enough memory.

**Comment by pcmoritz** [comment]
Hey, you will need to increase the size of the shared memory file system `/dev/shm` see http://www.walkernews.net/2010/05/04/how-to-resize-devshm-filesystem-in-linux/.

**Comment by pchalasani** [comment]
I think I need more than 67MB... the data-frames could be nearly 1 GB. And it complained that it cannot serialize a dataframe, so I used `pickle.dumps/loads`. 

Neptune.ml (like FloydHub) is a remote-GPU provider where I simply send a python script from the command line, and I don't have the ability to ssh to a specific box and change its `shm` settings.

If it were possible to do it via Python, I could put that at the top of the script. Is that possible?

**Comment by pchalasani** [comment]
I guess I could just run a system call from python

**Comment by pcmoritz** [comment]
Yeah, you can just call any shell command with the subprocess module.

The dataframes serialization was just added in https://github.com/ray-project/ray/commit/0684258d2ebd338d7bb564299052cd8dd44746c6 and sped up in https://github.com/ray-project/ray/commit/4157bcb80b169779f43a549dd503911d95507cc2. It is not part of the current release, we are creating a new point release in the next few days to include it.

There is also wheels for specific commits, like  
https://s3-us-west-2.amazonaws.com/ray-wheels/dc66a2d7d54e5e38055a600713535ccd40863b46/ray-0.2.1-cp36-cp36m-macosx_10_6_intel.whl, see the list at https://s3-us-west-2.amazonaws.com/ray-wheels/.

**Comment by pchalasani** [comment]
Come to think of it, this solution (of using my python script to edit the `/etc/fstab` to increase shared memory) seems very convoluted -- I mean surely, I'm not the first to encounter this problem of having to run ray on a remote machine. 

So I think I may be going about solving my problem the wrong way. I have a collection of (say) 30 files (in AWS/S3) that need to be read by my training class. I essentially have a loop:

```
for i in range(30):
   dataframe <- read file i (~ 20 secs)
   process dataframe (~ 30 secs)
```

Instead of doing this sequentially, I was hoping to leverage ray to asynchronously launch a collection `file_read()` jobs, and have a loop where after processing a dataframe, I grab the next "ready" dataframe (which would have been fetched while I was processing):

```
df_ids <- launch_readers()
for i in range(30):
    ready_id = get_next_ready_id()  # fast after the first one
    df <- ray.get(ready_id)  # fast
    process df    # ~ 30 secs
```

I first thought I would try to have a global cache (dict) that maps a file-index to the data-frame, and each file-reading task would add an entry to the dict: i -> data_frame_i, and the read task would simply return an index to the cache, rather than the actual data-frame, thus side-stepping the whole issue of shared memory size, serializability, etc.
But then this would need the tasks to be stateful, so it wouldn't work. I then thought about making my main class a remote class (actor), but I didn't like that because every method invocation (there are many) on that class would need to be made remote.

So is there some way I can maintain a global cache dict of index -> dataFrame, and use ray to asynchronously read files into data-frames and add them to the cache? Or some other paradigm I haven't thought of?

**Comment by pcmoritz** [comment]
Hey,

why not use the following scheme:

```
@ray.remote
def download_and_process_dataframe(s3_path):
    # load df from s3
    # process df
    # return processed df

cache = {s3_path: download_and_process_dataframe.remote(s3_path) for s3_path in s3_paths}

# can use the cache here to do other stuff like process the dataframe
```

It seems to me that in the example you gave, all the processing would happen serially on the driver.

If you want things to be more modular you can also have separate tasks to load and process the dataframes, in which case it would look something like this:

```
@ray.remote
def download_dataframe(s3_path):
    # load df from s3
    # return df

@ray.remote
def process_dataframe(df):
    # process df
    # return processed df

cache = {s3_path: download_dataframe.remote(s3_path) for s3_path in s3_paths}
cache = {s3_path: process_dataframe.remote(cache[s3_path]) for s3_path in s3_paths}
```

You will only have fun with Ray if the shared memory file system is large enough (say 40% of the total RAM). You should be able to set the size without messing with `fstab` by running

```
mount -o remount,size=8G /dev/shm
```

Here is how we run Ray on our clusters: http://ray.readthedocs.io/en/latest/using-ray-on-a-large-cluster.html

There is also instructions on how to do it more manually: http://ray.readthedocs.io/en/latest/using-ray-on-a-cluster.html

Let us know if you have more questions!

-- Philipp

**Comment by pchalasani** [comment]
Appreciate the detailed suggestions. I wondered whether it's possible to not have the remote function return the dataframe itself, but rather some index into a global cache variable that I maintain. This would sidestep the serialization issue plus the need to increase shared-mem.
It looks like this would not be possible, for some fundamental reason, namely that the tasks are supposed to be side-effect-free, am I right? 

Moreover, there is a serial dependency in the processing of my data-frames, so I cannot parallelize the processing itself; I just need the next data-frame to be ready when done processing the current one. This is why I was hoping to use Ray to fetch the data-frames asynchronously in the background.

Also, I tried the mount command above (as a python-subprocess command before doing `ray.init()`), but the remote machine gave a permission error, then I replaced the command with:

```
mount -o rw,uid=[my_user_id],remount,size=8G /dev/shm
```
using the following code...

```
        userid = getpass.getuser()
        process = subprocess.Popen('id ' + userid, shell=True, stdout=subprocess.PIPE)
        process.wait()
        fprint('resizing /dev/shm to 8G')
        command = 'mount -o rw,uid=' + userid + ',remount,size=8G /dev/shm'
        process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
        process.wait()
        fprint (process.returncode)
```
but now I get an error saying 

```
	mount: cannot remount shm read-write, is write-protected
```

**Comment by robertnishihara** [comment]
Ray requires a large shared-memory file system (on Linux it is often half the size of the total memory). In the absence of that it probably won't be possible to return large objects from remote functions or to pass large objects into remote functions.

As for the serialization issue, can you try `pip install -U ray`? We just pushed 0.2.2 to PyPI and that should improve the pandas integration.

Where do you want to store the dataframes (if not in the object store)? You could of course write them to disk. I'm not sure what you mean by "global cache variable that I maintain".

**Comment by pchalasani** [comment]
I meant I would keep a global in-memory dictionary variable `df_cache` that maps indices to data-frames. And I wanted to be able to write a ray remote fn that would read a file into a data-frame, save it in the cache at some index, and return that index (rather than the data-frame). So in effect the function would have a side-effect on this global `df_cache`. I guess that wouldn't be possible.

That said, I'm fine with re-sizing `/dev/shm` if I can get over the errors above.

And I don't really need to do `pip install -U ray` because I'm running jobs remotely on neptune.ml, and I have `ray` in requirements.txt file, and neptune will install the latest version.

**Comment by robertnishihara** [comment]
I see, the issue with the `df_cache` idea is that all of the workers run as separate Python processes so wouldn't naturally share global variables.

**Comment by pchalasani** [comment]
So that leaves resizing `/dev/shm` as the only option. I think this would actually be quite a common use-case for folks, especially as running DL/DRL jobs on remote GPUs becomes more common. So it would be great to have some instructions on how to go about resizing `/dev/shm` on a remote machine. Ideally, they should be some `subprocess.Popen(.. )` calls that we can put at the top of our script. Remember we cannot ssh to a specific box to do this. I've searched around quite a bit for how to handle permissions issues when doing this, but so far came up with nothing that works.

**Comment by pcmoritz** [comment]
Does `sudo mount -o remount,size=8G /dev/shm` work? You will probably need root privileges to create the FS, without it will be hard. Maybe you can ask the neptune.ml support? Another option would be to try FUSE and create a tmpfs, but I haven't tried that before.

**Comment by pchalasani** [comment]
That didn't work, that was the first thing I tried after suggested it. Yes I asked neptune support, and you're probably right about root privileges.

**Comment by pchalasani** [comment]
Turns out I was wrong about docker -- the neptune folks told me the jobs do run inside a docker container and I'm hoping they allow resizing shared memory

**Comment by robertnishihara** [comment]
Ok that makes sense. You typically have to resize it before starting the docker container, e.g., passing in `--shm-size`.

**Comment by pchalasani** [comment]
Yes I'm working with them to expose that option to me

**Comment by robertnishihara** [comment]
Closing this issue since I think we got to the bottom of it. Let us know if there are updates, and feel free to reopen.
